### ISSUE TITLE
Request: Powerful use-case for macros

### ISSUE THREAD

**Issue by rebcabin** [lcompilers/lpython#1972]
Issue #1908 is a plea for CTR in LPython in the distant future. CTR enables macros. However, even a special-purpose, simplified macro system à la [macropy](https://github.com/lihaoyi/macropy) would be immediately useful for cases like the following:

```python
    assert a_char == '(', "a_char == '('"
```

With even very simple macros, this could be shortened to 

```python
    assertm a_char == '('
```

**Comment by certik** [comment]
Or in this case we can simply improve our backend to print a much better error message for `assert a_char == '('` that includes the string itself, line number and a stacktrace.

**Comment by rebcabin** [comment]
wouldn't that be a language extension, thus violating one of our tenets? Macros can be achieved in ordinary CPython!

**Comment by Thirumalai-Shaktivel** [comment]
> Or in this case we can simply improve our backend to print a much better error message for `assert a_char == '('` that includes the string itself, line number and a stacktrace.

Yup, currently we have the following:
```console
$ lpython examples/expr2.py -g --debug-with-line-column
  File "examples/expr2.py", line 2
    assert t == 'c'
AssertionError
```

We can improve this or give it a try to use macros.

**Comment by certik** [comment]
The tenet is that if it runs in LPython, it will run in CPython and produce the same result. Our exceptions and failure modes do produce slightly different results (the stacktrace might look slightly differently, depending on how you compile it). I think we have some wiggle room there, and I thought a nicer assert error message would still fit the tenet.

The issue with macropy is that last commit was 6 years ago and I worry it will take the same route as the goto, that we eventually removed, because a new CPython version broke it. 

I feel macropy is actually extending Python, just like goto. It's extending Python with "hacks" that it was not meant to do. It works, until it gets broken by upstream. I would only commit to CPython as released upstream and its features. 

We do support somethings like structs and unions, but we do it via ctypes, which are supported upstream, so I think that's fine.

So I would not do macros, since Python doesn't have macros, so it would be against the tenets. But making an error message from assert slightly nicer I think would be fine and within tenets. :)

**Comment by certik** [comment]
Macros are powerful, but remember that AST changes from CPython version to version, and there are even some very slight differences with Python AST and LPython AST. So if we do it, we must tackle these challenges.

**Comment by rebcabin** [comment]
I totally agree! That's why macropy went out of maintenance! This is not a
small job, and for the far future if ever.

I'm NOT selling macros, in particular. I AM selling CTR (Conditional Term
Rewriting) as something we should take very seriously to propel LPython
into the world's BEST symbolic programming language SOMEDAY :)  Beyond
Mathematica.  My plea is in Issue #1908.

**Comment by certik** [comment]
Note that I thought #1908 is at the ASR level, while macropy is at the AST level. Rewriting ASR is powerful. The issue is that it's hard to do in pure Python mode without just calling LPython (but then you can't use debuggers). AST rewriting will work in both, but then it's tied to Python, unrelated to ASR.

**Comment by rebcabin** [comment]
yes, I wasn't all that clear about it. Rewriting applied to both AST and
ASR, but it's MUCH more interesting at ASR :)
